metadata-ws/src/main/java/uk/ac/ebi/ampt2d/metadata/persistence/services/TaxonomyServiceImpl.java

```diff
@@ -0,0 +1,80 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package uk.ac.ebi.ampt2d.metadata.persistence.services;
+
+import com.querydsl.jpa.impl.JPAQueryFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import uk.ac.ebi.ampt2d.metadata.persistence.entities.QTaxonomy;
+import uk.ac.ebi.ampt2d.metadata.persistence.entities.Taxonomy;
+import uk.ac.ebi.ampt2d.metadata.persistence.repositories.TaxonomyRepository;
+
+import javax.persistence.EntityManager;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.stream.Collectors;
+
+public class TaxonomyServiceImpl implements TaxonomyService {
+
+    @Autowired
+    private TaxonomyRepository taxonomyRepository;
+
+    @Autowired
+    private EntityManager entityManager;
+
+    @Override
+    public List<Long> findAllSpeciesAndSubspeciesTaxonomyIdsInATaxonomyTreeByTaxonomyId(long taxonomyId) {
+        QTaxonomy qTaxonomy = QTaxonomy.taxonomy;
+        QTaxonomy qTaxonomySpecies = new QTaxonomy("qTaxonomySpecies");
+        QTaxonomy qTaxonomyGenus = new QTaxonomy("qTaxonomyGenus");
+        QTaxonomy qTaxonomyOrder = new QTaxonomy("qTaxonomyOrder");
+        QTaxonomy qTaxonomyClass = new QTaxonomy("qTaxonomyClass");
+
+        JPAQueryFactory jpaQueryFactory = new JPAQueryFactory(entityManager);
+        List<Taxonomy> taxonomies = (List<Taxonomy>) jpaQueryFactory.from(qTaxonomy)
+                .leftJoin(qTaxonomy.taxonomySpecies, qTaxonomySpecies)
+                .leftJoin(qTaxonomy.taxonomyGenus, qTaxonomyGenus)
+                .leftJoin(qTaxonomy.taxonomyOrder, qTaxonomyOrder)
+                .leftJoin(qTaxonomy.taxonomyClass, qTaxonomyClass)
+                .where(qTaxonomySpecies.taxonomyId.eq(taxonomyId)
+                        .or(qTaxonomyOrder.taxonomyId.eq(taxonomyId)
+                                .or(qTaxonomyGenus.taxonomyId.eq(taxonomyId))
+                                .or(qTaxonomyClass.taxonomyId.eq(taxonomyId)))
+                        .or(qTaxonomy.taxonomyId.eq(taxonomyId))).fetch();
+        return getSpeciesAndSubspeciesTaxonomyIds(taxonomies);
+    }
+
+    @Override
+    public List<Long> findAllSpeciesAndSubspeciesTaxonomyIdsInATaxonomyTreeByTaxonomyName(String taxonomyName) {
+        Taxonomy taxonomy = taxonomyRepository.findByName(taxonomyName);
+        if (taxonomy == null) {
+            return new ArrayList<>();
+        }
+        return findAllSpeciesAndSubspeciesTaxonomyIdsInATaxonomyTreeByTaxonomyId(taxonomy.getTaxonomyId());
+    }
+
+    public List<Long> getSpeciesAndSubspeciesTaxonomyIds(List<Taxonomy> taxonomies) {
+        return taxonomies.parallelStream().filter(taxonomy -> {
+            String rank = taxonomy.getRank();
+            if (rank.equals("class") || rank.equals("order") || rank.equals("genus")) {
+                return false;
```

Comment:
I don't think there is need to perform this filtering at all, since entities such as samples and reference sequences won't ever contain these higher-rank taxonomies, and if _for some reason_ they do, then this needs to be displayed as well. Besides, this filtering requires additional computational time. It's simpler to just return all taxonomy IDs and then query  the required entities using them all